csrc/scheduler/normalization_utils.cpp

```diff
@@ -698,11 +702,61 @@ std::optional<int64_t> getOptionalInnerOuterPersistentBufferBatches(
       inner_batch,
       ", batch_min = ",
       batch_min);
+
   const int64_t batch_max = getMaximumInnerOuterPersistentBufferBatch();
-  if (inner_batch <= batch_max) {
-    return inner_batch;
+
+  // consider non-divisible splits.
+  // If inner_bath is still larger than bath_max and threads_per_block is
+  // smaller than threads_per_block_max possible. Try non-divisible batch
+  // splits.
+  while (inner_batch > batch_max &&
+         threads_per_block * 2 <= threads_per_block_max) {
+    threads_per_block *= 2;
+    inner_batch = ceilDiv(after_vectorization, threads_per_block);
+  }
+
+  // If inner_batch is still larger than batch_max, try set threads_per_block to
+  // max possible and recheck inner_batch, e.g. Initial after_vectorization =
+  // 9280/8 = 1160, threads_per_block = 8, inner_batch = 1160/8 = 145. After
+  // dividible split,  inner_batch = 29 and threads_per_block = 40. After
+  // non-divisible split, inner_batch = 8 and threads_per_block = 160. The
+  // following if statement will set threads_per_block to 256 and inner_batch
+  // to 5.
+  if (inner_batch > batch_max) {
+    threads_per_block = threads_per_block_max;
+    inner_batch = ceilDiv(after_vectorization, threads_per_block);
+  }
+
+  // If threads_per_block is smaller than threads_per_block_min, move
+  // parallelism from inner_batch to threads_per_block. e.g. Initial
+  // after_vectorization = 1344/8 = 168, threads_per_block = 8, inner_batch
+  // = 21. Up to there, threads_per_block = 24, inner_batch = 7. The following
+  // code will set threads_per_block to 128 and inner_batch to 2. And the
+  // latency for batch size 16K is reduced from 295 us to 150 us.
+  if (threads_per_block < threads_per_block_min) {
```

Comment:
Shouldn't `threads_per_block` be initialized to `threads_per_block_min` instead of `warp_size / 4`?

Reply:
Good point. In the revised strategy, we begin with `threads_per_block_min`. If the corresponding batch size exceeds `batch_max`, increment the threads per block by a warp. Continue this process until the block size reaches `threads_per_block_max` or the batch size reaches `batch_min` or the batch size is smaller or equal than `batch_max`.